drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,815 @@
+{
+	uint16_t data;
+
+	synopGMAC_read_phy_reg((u32 *)gmacdev->MacBase, addr, reg, &data);
+	return data;
+}
+
+static int numaker_eth_link_ok(synopGMACdevice *gmacdev)
+{
+	/* first, a dummy read to latch */
+	mdio_read(gmacdev, eth_phy_addr, MII_BMSR);
+	if (mdio_read(gmacdev, eth_phy_addr, MII_BMSR) & BMSR_LSTATUS) {
+		return 1;
+	}
+	return 0;
+}
+
+static int reset_phy(synopGMACdevice *gmacdev)
+{
+	uint16_t reg;
+	uint32_t delay_cnt;
+	int status = 0;
+
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, BMCR_RESET);
+
+	delay_cnt = 2000U;
+	while (delay_cnt > 0U) {
+		delay_cnt--;
+		if ((mdio_read(gmacdev, eth_phy_addr, MII_BMCR) & BMCR_RESET) == 0) {
+			break;
+		}
+	}
+
+	if (delay_cnt == 0U) {
+		LOG_DBG("Reset phy failed");
+		return -ENODATA;
+	}
+
+	delay_cnt = 200000U;
+	while (delay_cnt > 0) {
+		delay_cnt--;
+		if (numaker_eth_link_ok(gmacdev)) {
+			gmacdev->LinkState = LINKUP;
+			LOG_DBG("Link Up");
+			break;
+		}
+	}
+	if (delay_cnt == 0U) {
+		gmacdev->LinkState = LINKDOWN;
+		LOG_DBG("Link Down");
+		return -ENODATA;
+	}
+
+	mdio_write(gmacdev, eth_phy_addr, MII_ADVERTISE,
+		   ADVERTISE_CSMA | ADVERTISE_10HALF | ADVERTISE_10FULL | ADVERTISE_100HALF |
+			   ADVERTISE_100FULL);
+
+	reg = mdio_read(gmacdev, eth_phy_addr, MII_BMCR);
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, reg | BMCR_ANRESTART);
+
+	delay_cnt = 200000U;
+	while (delay_cnt > 0U) {
+		delay_cnt--;
+		if ((mdio_read(gmacdev, eth_phy_addr, MII_BMSR) &
+		     (BMSR_ANEGCOMPLETE | BMSR_LSTATUS)) == (BMSR_ANEGCOMPLETE | BMSR_LSTATUS)) {
+			break;
+		}
+	}
+
+	if (delay_cnt == 0U) {
+		LOG_DBG("AN failed. Set to 100 FULL");
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+		return -ENODATA;
+	}
+	reg = mdio_read(gmacdev, eth_phy_addr, MII_LPA);
+	if (reg & ADVERTISE_100FULL) {
+		LOG_DBG("100 full");
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_100HALF) {
+		LOG_DBG("100 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_10FULL) {
+		LOG_DBG("10 full");
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else {
+		LOG_DBG("10 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	}
+
+	printf("PHY ID 1:0x%x\r\n", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID1));
+	printf("PHY ID 2:0x%x\r\n", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID2));
```

Comment:
Please use the logging instead of adding printf()

Reply:
Yes, I changed as LOG_INF.